python/ray/experimental/state/api.py

```diff
@@ -1,195 +1,239 @@
-    r = requests.request(
-        "GET",
-        f"{api_server_url}/api/v0/{resource_name}",
-        params=params,
-        headers={"Content-Type": "application/json"},
-        json=None,
-        timeout=options.timeout,
-    )
-    r.raise_for_status()
+    def __init__(
+        self,
+        api_server_address: str = None,
+    ):
+        super().__init__(
+            self._get_default_api_server_address()
+            if api_server_address is None
+            else api_server_address,
+            create_cluster_if_needed=False,
+            headers={"Content-Type": "application/json"},
+        )
 
-    response = r.json()
-    if response["result"] is False:
-        raise RayStateApiException(
-            "API server internal error. See dashboard.log file for more details. "
-            f"Error: {response['msg']}"
+    @classmethod
+    def _get_default_api_server_address(cls) -> str:
+        assert (
+            ray.is_initialized()
+            and ray.worker.global_worker.node.address_info["webui_url"] is not None
         )
+        return f"http://{ray.worker.global_worker.node.address_info['webui_url']}"
+
+    @classmethod
+    def _get_query_string(cls, options: ListApiOptions) -> str:
+        query_strings = []
+        for field in fields(options):
+            query_strings.append(f"{field.name}={getattr(options, field.name)}")
+        return "&".join(query_strings)
+
+    def list(
+        self, resource: StateResource, options: ListApiOptions, _explain: bool = False
+    ) -> Union[Dict, List]:
+        """List resources states
+
+        Args:
+            resource_name: Resource names, i.e. 'jobs', 'actors', 'nodes',
+                see `StateResource` for details.
+            options: List options. See `ListApiOptions` for details.
+            _explain: Print the API information such as API
+                latency or failed query information.
+
+        Returns:
+            A list of queried result from `ListApiResponse`,
+
+        Raises:
+            This doesn't catch any exceptions raised when the underlying request
+            call raises exceptions. For example, it could raise `requests.Timeout`
+            when timeout occurs.
+
+        """
+        # Append the ListApiOptions to the url
+        endpoint = f"/api/v0/{resource.value}?{self._get_query_string(options)}"
+
+        # We don't use `asdict` to avoid deepcopy.
+        # https://docs.python.org/3/library/dataclasses.html#dataclasses.asdict
+        params = {
+            "limit": options.limit,
+            "timeout": options.timeout,
+            "filter_keys": [],
+            "filter_values": [],
+        }
+        for filter in options.filters:
+            filter_k, filter_val = filter
+            params["filter_keys"].append(filter_k)
+            params["filter_values"].append(filter_val)
+
+        # NOTE(rickyyx)
+        # This might raise `requests.Timeout` exceptions and other related requests
```

Comment:
Should we catch them and raise our built-in exception? (like StateApiRequestError)? Or alternatively, we can just mention in our API docstring that request lib's exceptions could be raised. Wdyt?

Reply:
Sure - we could. I thought It would be a direct catch and re-raise, which doesn't yield too much extra info. But now think about it, I could probably add some context in the exception (e.g. which endpoint is tried to talk to).